eclipse-theia/theia
# vscode: independent editor/title/run menu

#### What it does

In VS Code, contributions to the `"editor/title/run"` menu contribution point are not added to the ellipsis menu but to a dedicated item on the toolbar. This commit makes Theia do the same, except that unlike VS Code, a pop-up menu is always presented, even if there is only one action available.

In my opinion, the inconsistency of the VS Code implementation in not showing the Run/Debug pop-up menu but just the one contributed action on the toolbar, which looks quite different to the Run/Debug pop-up menu action, is confusing. If an action is contributed to an extension point that is a menu of Run/Debug actions, then that should show a menu of Run/Debug actions even if there is only one item on the menu.

_If it really is necessary to emulate this detail of the VS Code behaviour, then that should be a follow-up issue to discuss with the user community._

Note also that this PR uses the `debug-alt` icon from the VS Code `codicons` set as the best available alternative from that set to the icon that VS Code uses for its "Run/Debug" menu, which is not in the `codicons` set. If somebody would like to contribute a better license-compatible open-source icon for the purpose, please do!

Fixes #12687

#### How to test

1. Clone the `menu_bug_3` of the OP's sample repository used for other recent bug reports: [Theia Sandbox](/jcortell68/theiasandbox).
2. Change the menu contribution in the `myext` extension to the `"editor/title/run"` point (it should be checked out as `"editor/lineNumber/context"` for another issue report) and remove the `"group"` property that isn't needed for this use case.
3. Because there probably wouldn't be any other actions on the "more" menu, I recommend adding the `"myext.goodbyeWorld"` command from this extension to the `"editor/title"` menu so that we can see that this menu exists and doesn't get contributions intended for the `"editor/title/run"` menu.
4. Build the `myext` extension and link it into your `plugins/` directory.
5. Launch Theia and open some editor (e.g., TS or JSON file).
6. Observe that the "Run/Debug" menu shows up something like the screen recording below, with the "Hello World" action from the extension. Additionally, hovering the pointer over it shows a "Run or Debug..." tooltip, similar to the "More Items..." tooltip on the `...` "more" menu.
7. Verify that the "Hello World" action is _not_ in the `...` menu.

Also worth verifying are some related use cases:

- check that a Theia instance without any contributions to the `"editor/title/run"` menu does not show that menu in the toolbar.
- also that in the case there are contributions to this menu, but all of them are suppressed by their `"when"` clauses not matching the current context, then in this case also the menu does not appear in the toolbar.

![CleanShot 2023-08-04 at 13 38 43](https://github.com/eclipse-theia/theia/assets/1615558/2a54ed73-32f3-4fae-a300-8d8fac5d38db)

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
